drivers/console/uart_console.c

```diff
@@ -335,6 +339,114 @@ static void handle_ansi(u8_t byte, char *line)
 	atomic_clear_bit(&esc_state, ESC_ANSI);
 }
 
+#ifdef CONFIG_UART_CONSOLE_MCUMGR
+
+static void clear_mcumgr(void)
+{
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_PKT);
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_PKT_2);
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_DATA);
+	atomic_clear_bit(&esc_state, ESC_MCUMGR_DATA_2);
+}
+
+/**
+ * These states indicate whether an mcumgr frame is being received.
+ */
+#define CONSOLE_MCUMGR_STATE_NONE       1
+#define CONSOLE_MCUMGR_STATE_HEADER     2
+#define CONSOLE_MCUMGR_STATE_PAYLOAD    3
+
+static int read_mcumgr_byte(uint8_t byte)
+{
+	bool data_1;
+	bool data_2;
+	bool pkt_1;
+	bool pkt_2;
+
+	pkt_1 = atomic_test_bit(&esc_state, ESC_MCUMGR_PKT);
+	pkt_2 = atomic_test_bit(&esc_state, ESC_MCUMGR_PKT_2);
+	data_1 = atomic_test_bit(&esc_state, ESC_MCUMGR_DATA);
+	data_2 = atomic_test_bit(&esc_state, ESC_MCUMGR_DATA_2);
+
+	if (pkt_2 || data_2) {
+		/* Already fully framed. */
+		return CONSOLE_MCUMGR_STATE_PAYLOAD;
+	}
+
+	if (pkt_1) {
+		if (byte == MCUMGR_SERIAL_HDR_PKT_2) {
+			/* Final framing byte received. */
+			atomic_set_bit(&esc_state, ESC_MCUMGR_PKT_2);
+			return CONSOLE_MCUMGR_STATE_PAYLOAD;
+		}
+	} else if (data_1) {
+		if (byte == MCUMGR_SERIAL_HDR_FRAG_2) {
+			/* Final framing byte received. */
+			atomic_set_bit(&esc_state, ESC_MCUMGR_DATA_2);
+			return CONSOLE_MCUMGR_STATE_PAYLOAD;
+		}
+	} else {
+		clear_mcumgr();
+		if (byte == MCUMGR_SERIAL_HDR_PKT_1) {
+			/* First framing byte received. */
+			atomic_set_bit(&esc_state, ESC_MCUMGR_PKT);
+			return CONSOLE_MCUMGR_STATE_HEADER;
+		} else if (byte == MCUMGR_SERIAL_HDR_FRAG_1) {
+			/* First framing byte received. */
+			atomic_set_bit(&esc_state, ESC_MCUMGR_DATA);
+			return CONSOLE_MCUMGR_STATE_HEADER;
+		}
+	}
+
+	/* Non-mcumgr byte received. */
+	return CONSOLE_MCUMGR_STATE_NONE;
+}
+
+/**
+ * @brief Attempts to process a received byte as part of an mcumgr frame.
+ *
+ * @param cmd The console command currently being received.
+ * @param byte The byte just received.
+ *
+ * @return true if the command being received is an mcumgr frame; false if it
+ * is a plain console command.
+ */
+static bool handle_mcumgr(struct console_input *cmd, uint8_t byte)
+{
+	int mcumgr_state;
+
+	mcumgr_state = read_mcumgr_byte(byte);
+	if (mcumgr_state == CONSOLE_MCUMGR_STATE_NONE) {
+		/* Not an mcumgr command; let the normal console handling
+		 * process the byte.
+		 */
+		cmd->is_mcumgr = 0;
+		return false;
+	}
+
+	/* The received byte is part of an mcumgr command.  Process the byte
+	 * and return true to indicate that normal console handling should
+	 * ignore it.
+	 */
+	if (cur + end < sizeof(cmd->line) - 1) {
+		cmd->line[cur++] = byte;
+	}
+	if (mcumgr_state == CONSOLE_MCUMGR_STATE_PAYLOAD && byte == '\n') {
```

Comment:
Are we doing all this processing for every byte that comes over UART in a DFU payload?

Perhaps we should do some some sort of benchmarking to make sure this implementations
 performance will allow this feature to be used for it's intended use-case.

Reply:
The code gets executed for each received byte, but I don't think it is that much processing.  Perhaps I'm being naive, but my sense is that the UART delivers data too slowly for this to be a problem.

Reply:
On nRF52832 UART can do 1Mbps straight to RAM while the CPU is sleeping 255 bytes at a time (after 255 bytes the CPU needs to wake up and update the DMA pointer).

IMHO, we can allow mcumgr to provide 10% / 5 % of the HW-capable DFU-speed if and only if we are confident that reaching the potential DFU speed will not require userspace-breaking changes to the mcumgr API.

In other words, if you think we can improve the implementation later without affecting users then this is fine. If not, we need to get this right the first time.

PS: I guess, in most serial DFU-situations, it will always be the flash that is the bottle-neck. For nrf52832 we have 17us per byte flash write, whereas UART 1Mbps gives 8us per byte flash write. The point still stands though. We need to be confident that mcumgr doesn't need to be re-written to support the performance potential of serial DFU.